Add JSON casts to range types.

vpetrovykh opened the pull request:
Add casts to and from JSON to all our range types. The JSON generated is
structured as an object with the following fields: "lower", "inc_lower",
"upper", "inc_upper".

1st1 commented on tests/test_edgeql_expressions.py:
```diff
@@ -5941,6 +5941,238 @@ async def test_edgeql_expr_range_31(self):
                     )
                 ''')
 
+    async def test_edgeql_expr_range_32(self):
```
I don't see any error cases. What happens if I pass various kinds of garbage to `select <range<cal::local_date>>to_json($1)`? Think missing some attributes, or an outright wrong kind of JSON (array, number 42, object literal with all wrong keys, etc)

vpetrovykh replied:
I've added a bunch more tests for various wrong JSON inputs.

1st1 reviewed (changes requested):
Needs tests for error conditions.

vpetrovykh commented:
I've fixed the issue mentioned by @msullivan with json mode (where ranges nested in shapes would not cast to JSON properly) and added a test with a nested shape.

elprans commented on edb/pgsql/compiler/output.py:
```diff
@@ -557,6 +557,14 @@ def serialize_expr_to_json(
         )
 
+    elif irtyputils.is_range(styperef) and not expr.ser_safe:
+        cast_name = s_casts.get_cast_fullname_from_names(
```
What's the point of getting the cast_name then?

vpetrovykh replied:
Oops, no point. Fixed.

elprans reviewed (approved):
LGTM basically, but let's add serialization support to remove the need for `json_only` hack and have proper output tests
